privates/widgets.py

```diff
@@ -16,10 +17,20 @@ def get_context(
     ):
         """
         Return value-related substitutions.
+
+        If a view exists for `/url_name/<pk>`, add the url to the context and mark
+        the corresponding file as apt for download. Otherwise, add empty url.
         """
         context = super().get_context(name, value, attrs)
         if self.is_initial(value):
-            context["url"] = reverse(self.url_name, kwargs={"pk": value.instance.pk})
+            try:
+                context["url"] = reverse(
+                    self.url_name, kwargs={"pk": value.instance.pk}
+                )
+                context["download_permitted"] = True
+            except NoReverseMatch:
+                context["url"] = ""
+                context["download_permitted"] = False
```

Comment:
in `admin.PrivateMediaMixin.formfield_for_dbfield` the `Widget` is instantiated on the fly, you can pass and extra kwarg `download_allowed` there based on if `db_field.name not in self.private_media_no_download_fields`.

Next, extend `widgets.PrivateFileWidgetMixin.__init__` to take an additional kwarg `download_allowed`, and then you can simply check it with if/else instead of the `try: except` approach.